packages/langium/src/grammar/langium-grammar.langium

```diff
@@ -135,40 +132,43 @@ ParenthesizedElement returns AbstractElement:
 PredicatedGroup returns Group:
 	(predicated?='=>' | firstSetPredicated?='->') '(' elements+=Alternatives ')';
 
-TerminalRule :
-	'terminal' (^fragment?='fragment' name=ID | name=ID ('returns' type=ID)?) ':'
-		regex=RegexLiteral
+TerminalRule:
+	^hidden?='hidden'? 'terminal' (^fragment?='fragment' name=ID | name=ID ('returns' type=ID)?) ':'
+		^terminal=TerminalAlternatives
 	';';
 
 terminal RegexLiteral returns string: /\/(?![*+?])(?:[^\r\n\[/\\]|\\.|\[(?:[^\r\n\]\\]|\\.)*\])+\//;
 
-TerminalAlternatives:
+TerminalAlternatives returns AbstractElement:
 	TerminalGroup ({TerminalAlternatives.elements+=current} '|' elements+=TerminalGroup)*;
 
-TerminalGroup:
+TerminalGroup returns AbstractElement:
 	TerminalToken ({TerminalGroup.elements+=current} elements+=TerminalToken+)?;
 
-TerminalToken:
+TerminalToken returns AbstractElement:
 	TerminalTokenElement cardinality=('?'|'*'|'+')?;
 
-TerminalTokenElement:
-	CharacterRange | TerminalRuleCall | ParenthesizedTerminalElement | AbstractNegatedToken | Wildcard;
+TerminalTokenElement returns AbstractElement:
+	CharacterRange | TerminalRuleCall | ParenthesizedTerminalElement | NegatedToken | UntilToken | RegexToken | Wildcard;
 
-ParenthesizedTerminalElement returns TerminalAlternatives:
+ParenthesizedTerminalElement returns AbstractElement:
 	'(' TerminalAlternatives ')';
 
-AbstractNegatedToken:
-	NegatedToken | UntilToken;
+TerminalRuleCall returns AbstractElement:
+	{TerminalRuleCall} rule=[TerminalRule|ID];
+
+NegatedToken returns AbstractElement:
+	{NegatedToken} '!' ^terminal=TerminalTokenElement;
 
-NegatedToken:
-	'!' ^terminal=TerminalTokenElement;
+UntilToken returns AbstractElement:
+	{UntilToken} '->' ^terminal=TerminalTokenElement;
 
-UntilToken:
-	'->' ^terminal=TerminalTokenElement;
+RegexToken returns AbstractElement:
+	{RegexToken} regex=RegexLiteral;
 
-Wildcard:
+Wildcard returns AbstractElement:
 	{Wildcard} '.';
 
-CharacterRange:
-	left=Keyword ('..' right=Keyword)?;
+CharacterRange returns AbstractElement:
+	{CharacterRange} left=Keyword ('..' right=Keyword)?;
```

Comment:
Given that we can express the same with regular expressions, some of these constructs are redundant. We should still keep them to ease migration from Xtext. But what do you think about adding hints through validation to discourage some constructs in favor of the regexp equivalent?

Reply:
Added a hint and a quick-fix for that 👍